### PR TITLE
[BEAM-11142] Enable CrossLanguageValidateRunner test for dataflow run…

### PR DESCRIPTION
**Pull request by ihji**
…ner v2

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ihji** [comment]
run seed job

**Comment by ihji** [comment]
run XVR_flink postcommit

**Comment by ihji** [comment]
run XVR_dataflow postcommit

**Comment by ihji** [comment]
run seed job

**Comment by ihji** [comment]
run xvr_flink postcommit

**Comment by ihji** [comment]
run xvr_spark postcommit

**Comment by ihji** [comment]
run xvr_direct postcommit

**Comment by ihji** [comment]
run xvr_dataflow postcommit

**Comment by ihji** [comment]
R: @chamikaramj @robertwb

File: sdks/python/apache_beam/transforms/ptransform.py [hunk 1]
```diff
@@ -695,9 +695,9 @@ def from_runner_api(cls,
     # type: (...) -> Optional[PTransform]
     if proto is None or proto.spec is None or not proto.spec.urn:
       return None
-    parameter_type, constructor = cls._known_urns[proto.spec.urn]
 
     try:
+      parameter_type, constructor = cls._known_urns[proto.spec.urn]
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/ptransform.py:700]: Ditto. Not sure why we need this update given that tests should already pass for Dataflow.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 2]
```diff
@@ -542,19 +593,32 @@ def run_pipeline(self, pipeline, options):
...
       # leaf elements.
       pipeline.visit(self._set_pdone_visitor(pipeline))
 
+      from apache_beam.runners import pipeline_context
+      dst_context = pipeline_context.PipelineContext(
+          component_id_map=pipeline.component_id_map,
+          default_environment=self._default_environment)
+
+      # Copy external environments to prevent dangling environment ids
+      pipeline.visit(
```
Comment (chamikaramj) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:613]: Why do we need to update dataflow_runner for the test ?
Existing tests should already work for dataflow_runner without these updates.

Duplicates environments are handled here when submitting the Dataflow job request: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/dataflow/internal/apiclient.py#L303

File: sdks/python/setup.py [hunk 3]
```diff
@@ -256,70 +256,71 @@ def run(self):
...
-        ]},
-    cmdclass={
-        'build_py': generate_protos_first(build_py),
-        'develop': generate_protos_first(develop),
-        'egg_info': generate_protos_first(egg_info),
-        'test': generate_protos_first(test),
-        'mypy': generate_protos_first(mypy),
-    },
-)
+if __name__ == '__main__':
```
Comment (chamikaramj) [inline sdks/python/setup.py:259]: This probably should be a separate PR.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 2 above)
Comment (ihji) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:613]: Python pipeline using python external transform doesn't work without handling dangling environment ids.

It's not because of duplicated environments. It's because of missing environments. Dangling environment IDs are generated when 1) PTransform URN from external environment is known (therefore `RunnerAPIPTransformHolder` won't work as expected) 2) PipelineContext from rehydration process is not used in the second proto conversion. Python external transforms in Python pipelines are rehydrated to `AppliedPTransform` since external PTransform URN is known to the SDK however AppliedPTransform only keeps an environment id not environment itself. Environments are saved separately in PipelineContext but we dropped the PipelineContext from rehydration in the second proto conversion.

File: sdks/python/apache_beam/transforms/ptransform.py (hunk 1 above)
Comment (ihji) [reply sdks/python/apache_beam/transforms/ptransform.py:700]: It was for fixing Java xlang pipelines but the fix is still incomplete anyway. Will revert this.

**Comment by TheNeuralBit** [comment]
This should run the SQL tests as well since they are added as part of createCrossLanguageValidatesRunnerTask right?

https://github.com/apache/beam/blob/88d1696b6a4bb0b98eb289e9118ac4064127378e/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L1980

If so this will also address BEAM-10011

File: sdks/python/setup.py (hunk 3 above)
Comment (ihji) [reply sdks/python/setup.py]: Done.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 2 above)
Comment (chamikaramj) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:613]: "RunnerAPIPTransformHolder" is expected to go away soon (when we enable portable job submission). Do you think we can skip this test (that use Python external transforms in a Python pipeline) for now and try without these updates when we have portable job submission ?

**Comment by ihji** [comment]
run seed job

**Comment by ihji** [comment]
run xvr_dataflow postcommit

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 2 above)
Comment (ihji) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Make sense. Updated to enable PythonUsingJava and PythonUsingSql only.

File: runners/google-cloud-dataflow-java/build.gradle [hunk 4]
```diff
@@ -312,6 +313,37 @@ task validatesRunnerStreaming {
...
+  classpath: configurations.validatesRunner,
+  numParallelTests: Integer.MAX_VALUE,
+  needsSdkLocation: true,
+  pythonPipelineOptions: [
+    "--runner=TestDataflowRunner",
+    "--project=${dataflowProject}",
+    "--region=${dataflowRegion}",
+    "--sdk_harness_container_image_overrides=.*java.*,${dockerImageContainer}:${dockerTag}",
+    "--experiments=beam_fn_api",
+    "--experiments=use_unified_worker",
```
Comment (chamikaramj) [inline runners/google-cloud-dataflow-java/build.gradle:328]: Please remove experiments "beam_fn_api" and "unified_worker" and add "use_runner_v2"

File: runners/google-cloud-dataflow-java/build.gradle [hunk 5]
```diff
@@ -312,6 +313,37 @@ task validatesRunnerStreaming {
...
+  ],
+  javaPipelineOptions: [
+    "--runner=TestDataflowRunner",
+    "--project=${dataflowProject}",
+    "--region=${dataflowRegion}",
+    "--tempRoot=${dataflowValidatesTempRoot}",
+    //"--sdkHarnessContainerImageOverrides=.*java.*,${dockerImageContainer}:${dockerTag}",
+    "--experiments=beam_fn_api,use_unified_worker",
```
Comment (chamikaramj) [inline runners/google-cloud-dataflow-java/build.gradle:336]: Ditto.

File: .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+          switches('-PpythonVersion=3.6')
+        }
+        shell('echo "*** RUN CROSS-LANGUAGE DATAFLOW PYTHON WITH JAVA EXTERNAL TRANSFORMS USING PYTHON 3.8 ***"')
+        gradle {
+          rootBuildScriptDir(commonJobProperties.checkoutDir)
+          tasks(':runners:google-cloud-dataflow-java:validatesCrossLanguageRunnerPythonUsingJava')
+          commonJobProperties.setGradleSwitches(delegate)
+          switches('-PpythonVersion=3.8')
+        }
+        shell('echo "*** RUN CROSS-LANGUAGE DATAFLOW PYTHON WITH JAVA SQL TRANSFORMS USING PYTHON 3.8 ***"')
```
Comment (chamikaramj) [inline .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy:54]: Can we add Kafka as well ?

**Review by chamikaramj — COMMENTED** [review]
Thanks!

File: runners/google-cloud-dataflow-java/build.gradle (hunk 4 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/build.gradle]: done.

File: runners/google-cloud-dataflow-java/build.gradle (hunk 5 above)
Comment (ihji) [reply runners/google-cloud-dataflow-java/build.gradle]: done.

**Comment by ihji** [comment]
run seed job

File: .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy (hunk 6 above)
Comment (ihji) [reply .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy:54]: I think it's not necessary. Kafka xlang test is already a part of python postcommit tests: https://github.com/apache/beam/blob/master/sdks/python/test-suites/portable/common.gradle#L177

**Review by chamikaramj — APPROVED** [review]
LGTM. Thanks.

**Comment by ihji** [comment]
run xvr_dataflow postcommit

**Comment by ihji** [comment]
Run Java PreCommit

**Comment by ihji** [comment]
Run Java_Examples_Dataflow PreCommit

File: .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy (hunk 6 above)
Comment (TheNeuralBit) [reply .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy:54]: That only runs Kafka on Flink though right?

and thank you for adding the SQL task :) it looks like it's working:
image

Comment (ihji) [reply .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy:54]: Oh, that's right. Good catch 😄
Will take a look if we can easily add Kafka xlang test too.

Comment (ihji) [reply .test-infra/jenkins/job_PostCommit_CrossLanguageValidatesRunner_Dataflow.groovy:54]: We can't simply adding xlang Kafkaio test here since it needs a separate Kafka cluster setup on Dataflow (on Flink it uses locally launched Kafka cluster). Will merge without xlang Kafkaio test.

**Comment by ihji** [comment]
Run Python PreCommit

File: runners/google-cloud-dataflow-java/build.gradle [hunk 7]
```diff
@@ -312,6 +313,36 @@ task validatesRunnerStreaming {
   ))
 }
 
+createCrossLanguageValidatesRunnerTask(
```
Comment (boyuanzz) [inline runners/google-cloud-dataflow-java/build.gradle:316]: This task makes all Dataflow runner v2 tests fail because the sdk container is cleaned up as soon as it is built. The cause is that in the common configuration, you have
config.startJobServer.finalizedBy config.cleanupJobServer
This will change the task dependency incorrectly.

Filed jira here: https://issues.apache.org/jira/browse/BEAM-11270

cc: @tysonjh
